kore/src/main/scala/org/kframework/definition/outer-ext.scala

```diff
@@ -57,90 +24,44 @@ case class Bubble(sentenceType: String, contents: String, att: Att = Att.empty)
 }
 
-object FlatModule {
-  def apply(name: String, unresolvedLocalSentences: Set[Sentence]): FlatModule = {
-    new FlatModule(name, Set(), unresolvedLocalSentences, Att.empty)
-  }
-}
-
 case class Import(name: String, att: Att = Att.empty) extends HasLocation {
   override def location(): Optional[Location] = att.getOptional(classOf[Location])
   override def source(): Optional[Source] = att.getOptional(classOf[Source])
 }
 
-object Import {
-  val syntaxString = "$SYNTAX"
-
-  def isSyntax(name: String): Boolean = name.endsWith(syntaxString)
-
-  def asSyntax(_import: Import): Import =
-    if (isSyntax(_import.name))
-      _import
-    else
-      Import(_import.name ++ syntaxString, _import.att)
-
-  def noSyntax(name: String): String =
-    if (isSyntax(name))
-      name.dropRight(syntaxString.length)
-    else
-      name
-}
-
-
-
 case class FlatModule(name: String, imports: Set[Import], localSentences: Set[Sentence], @(Nonnull@param) val att: Att = Att.empty)
   extends OuterKORE with Sorting with Serializable {
+}
 
-  def toModule(allModules: Set[FlatModule], koreModules: scala.collection.mutable.Map[String, Module], visitedModules: Seq[FlatModule]): Module = {
-    var items = this.localSentences
-    var importedModuleNames = this.imports
-    var importedSyntax = importedModuleNames.map(m => Import.asSyntax(m))
-
-    if (visitedModules.contains(this)) {
-      var msg = "Found circularity in module imports: "
-      visitedModules.foreach(m => msg += m.name + " < ")
-      msg += visitedModules.head.name
-      throw KEMException.compilerError(msg)
-    }
-
-    if (koreModules.contains(this.name))
-      return koreModules(this.name)
+object FlatModule {
+  def apply(name: String, unresolvedLocalSentences: Set[Sentence]): FlatModule = {
+    new FlatModule(name, Set(), unresolvedLocalSentences, Att.empty)
+  }
 
-    def resolveImport(_import: Import): Module = {
-      var baseName = Import.noSyntax(_import.name)
-      var modOption = allModules.find(m => m.name.equals(baseName))
-      if (modOption.nonEmpty) {
-        var mod = modOption.get
-        var result = koreModules.get(mod.name)
-        if (result.isEmpty) {
-          result = Some(mod.toModule(allModules, koreModules, this +: visitedModules))
-        }
-        if (Import.isSyntax(_import.name)) {
-          result = Some(koreModules.get(_import.name).get)
-        }
-          result.get
-      } else if (koreModules.contains(_import.name))
-          koreModules.get(_import.name).get
-        else
-          throw KEMException.compilerError("Could not find module: " + _import.name, _import)
+  def toModule(allModules:Seq[FlatModule], previousModules:Set[Module]):Set[Module] = {
+    val memoization:TrieMap[String, Module] = collection.concurrent.TrieMap[String, Module]()
+    previousModules.map(m => memoization.put(m.name, m))
+    allModules.map(m => toModuleRec(m, Seq()))
+    def toModuleRec(m:FlatModule, visitedModules: Seq[FlatModule]):Module = {
```

Comment:
It's a little confusing to me to use `toModuleRec` here on the line before it's declared. Can we just do the `allModules.map ...` line after the declaration of `toModuleRec`?